File: experiments/10-runtimes/src/node/churn.mjs
// E2, partie 3 : allocation continue à ensemble vivant fixe en JavaScript (Node, V8).
// Même protocole que src/c/churn.c ; collectes observées par PerformanceObserver.

import { readFileSync, writeFileSync } from "node:fs";
import { PerformanceObserver } from "node:perf_hooks";

const opt = { "--live-log2": "20", "--batch-log2": "13", "--batches": "1024" };
for (let i = 2; i + 1 < process.argv.length; i += 2) opt[process.argv[i]] = process.argv[i + 1];
const live = 2 ** Number(opt["--live-log2"]);
const batch = 2 ** Number(opt["--batch-log2"]);
const batches = Number(opt["--batches"]);

let gcCount = 0;
let gcDuration = 0;
let measuring = false;
const observer = new PerformanceObserver((list) => {
  for (const entry of list.getEntries()) {
    if (measuring) {
      gcCount += 1;
      gcDuration += entry.duration;
    }
  }
});
observer.observe({ entryTypes: ["gc"] });

function make(i) {
  return { key: i, a: 2 * i + 1, b: 3 * i + 2 };
}

const ring = new Array(live);
let i = 0;
for (; i < live; i++) ring[i] = make(i);

// Les entrées de collecte sont livrées de façon asynchrone : on laisse passer celles du
// remplissage avant d'ouvrir la fenêtre, et celles de la mesure avant de la fermer.
await new Promise((resolve) => setTimeout(resolve, 50));
measuring = true;
const elapsed = new Array(batches);
const start = process.hrtime.bigint();
for (let k = 0; k < batches; k++) {
  const t0 = process.hrtime.bigint();
  for (let j = 0; j < batch; j++) {
    ring[i % live] = make(i);
    i++;
  }
  elapsed[k] = process.hrtime.bigint() - t0;
}
const total = process.hrtime.bigint() - start;
await new Promise((resolve) => setTimeout(resolve, 50));
measuring = false;
observer.disconnect();

let checksum = 0;
for (const o of ring) checksum += o.key + o.a + o.b;
const status = readFileSync("/proc/self/status", "utf8");
const rss = Number((status.match(/VmHWM:\s+(\d+)/) ?? [0, -1])[1]);
const allowedCpus = (readFileSync("/proc/self/status", "utf8").match(/Cpus_allowed_list:\s+(\S+)/) ?? [0, "inconnu"])[1].replaceAll(",", ";");
const lines = ["run_id,mode,batch,operations,elapsed_ns"];
for (let k = 0; k < batches; k++) lines.push(`${opt["--run-id"]},${opt["--mode"]},${k},${batch},${elapsed[k]}`);
writeFileSync(opt["--output"], lines.join("\n") + "\n");
writeFileSync(opt["--meta"], "run_id,mode,runtime,total_ns,operations,checksum,gc_count,gc_pause_ns,peak_rss_kib,cpus_allowed\n" +
  `${opt["--run-id"]},${opt["--mode"]},node ${process.version},${total},${batches * batch},${checksum},${gcCount},` +
  `${Math.round(gcDuration * 1e6)},${rss},${allowedCpus}\n`);

File: experiments/10-runtimes/src/java/Churn.java
// E2, partie 3 : allocation continue à ensemble vivant fixe en Java. Même protocole que
// src/c/churn.c ; le collecteur est choisi par option de la seule JVM (G1 ou Serial).

import java.io.IOException;
import java.io.PrintWriter;
import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.ManagementFactory;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;

public final class Churn {
    static final class Obj {
        final long key;
        final long a;
        final long b;

        Obj(long i) {
            key = i;
            a = 2 * i + 1;
            b = 3 * i + 2;
        }
    }

    static long[] gcTotals() {
        long count = 0;
        long millis = 0;
        for (GarbageCollectorMXBean bean : ManagementFactory.getGarbageCollectorMXBeans()) {
            count += Math.max(0, bean.getCollectionCount());
            millis += Math.max(0, bean.getCollectionTime());
        }
        return new long[] {count, millis};
    }

    static long peakRssKib() throws IOException {
        for (String line : Files.readAllLines(Path.of("/proc/self/status"))) {
            if (line.startsWith("VmHWM:")) {
                return Long.parseLong(line.split("\\s+")[1]);
            }
        }
        return -1;
    }

    static String allowedCpus() throws IOException {
        for (String line : Files.readAllLines(Path.of("/proc/self/status"))) {
            if (line.startsWith("Cpus_allowed_list:")) {
                return line.substring(line.indexOf(':') + 1).trim().replace(',', ';');
            }
        }
        return "inconnu";
    }

    public static void main(String[] args) throws IOException {
        HashMap<String, String> opt = new HashMap<>();
        for (int i = 0; i + 1 < args.length; i += 2) {
            opt.put(args[i], args[i + 1]);
        }
        int live = 1 << Integer.parseInt(opt.getOrDefault("--live-log2", "20"));
        int batch = 1 << Integer.parseInt(opt.getOrDefault("--batch-log2", "13"));
        int batches = Integer.parseInt(opt.getOrDefault("--batches", "1024"));
        String runId = opt.get("--run-id");
        String mode = opt.get("--mode");

        Obj[] ring = new Obj[live];
        long i = 0;
        for (; i < live; i++) {
            ring[(int) i] = new Obj(i);
        }
        long[] elapsed = new long[batches];
        long[] before = gcTotals();
        long start = System.nanoTime();
        for (int k = 0; k < batches; k++) {
            long t0 = System.nanoTime();
            for (int j = 0; j < batch; j++) {
                ring[(int) (i & (live - 1))] = new Obj(i);
                i++;
            }
            elapsed[k] = System.nanoTime() - t0;
        }
        long total = System.nanoTime() - start;
        long[] after = gcTotals();
        long checksum = 0;
        for (Obj o : ring) {
            checksum += o.key + o.a + o.b;
        }

        try (PrintWriter out = new PrintWriter(opt.get("--output"))) {
            out.println("run_id,mode,batch,operations,elapsed_ns");
            for (int k = 0; k < batches; k++) {
                out.printf("%s,%s,%d,%d,%d%n", runId, mode, k, batch, elapsed[k]);
            }
        }
        String collectors = String.join("+", ManagementFactory.getGarbageCollectorMXBeans().stream()
                .map(GarbageCollectorMXBean::getName).map(n -> n.replace(' ', '_')).toList());
        try (PrintWriter meta = new PrintWriter(opt.get("--meta"))) {
            meta.println("run_id,mode,runtime,total_ns,operations,checksum,gc_count,gc_pause_ns,peak_rss_kib,cpus_allowed");
            meta.printf("%s,%s,java %s %s,%d,%d,%s,%d,%d,%d,%s%n", runId, mode, System.getProperty("java.vm.version"),
                    collectors, total, (long) batches * batch, Long.toUnsignedString(checksum), after[0] - before[0],
                    (after[1] - before[1]) * 1_000_000L, peakRssKib(), allowedCpus());
        }
    }
}
